records/2025-04-14.batch.saved.ses.receipts.md

```diff
@@ -0,0 +1,80 @@
+
+## Options
+
+There's two main options that we have arrived at to solve this at the **lambda level**:
+* Use Redis as a buffer queue to store and flush receipts once a specified `BATCH_SIZE` of receipts have accumulated. Involves both app code and infrastructure changes.
+* Add a new SQS queue to buffer receipts. The queue will sit between the SNS topic and the callback lambda. This option is primarily an infrastructure change.
+
+### Option 1 - Introduce a new SQS queue to act as a buffer queue
+Unlike SNS, subscribers to SQS queues can be configured to pick up batches of messages at a time from the queue. In the current implementation:
+
+1. SES puts a receipt into an SNS topic
+2. The callback lambda, which is subscribed to this topic, pulls 1 receipt event at a time
+3. The message is repackaged into a celery task and sent to `process_ses_results` to update the DB, perform API callbacks, etc.
+
+![current implementation](./diagrams/2025-04-14.batch.saved.ses.receipts/current-implementation.png)
+
+To enable our callback lambda to receive batches of receipts, a new SQS queue is introduced sitting between the SNS topic and lambda, and subscriptions are re-arranged:
+
+1. The lambda subscribes to the new `SQS buffer queue` instead of the SNS topic
+2. The `SQS buffer queue` subscribes to the SNS topic
+
+![current implementation](./diagrams/2025-04-14.batch.saved.ses.receipts/proposed-added-sqs-queue.png)
+
+To achieve the desired batching behaviour, the lambda's subscription to the new SQS buffer queue is configured with:
+- A [BATCH_SIZE](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping#batch_size-1) indicating how many messages to transmit when calling the lambda
+- A [`MAX_BATCHING_WINDOW`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping#maximum_batching_window_in_seconds-1) indicating how long to wait between calls to the lambda
+
+The SQS queue continually receives receipt events from the SNS topic and flushes them to the lambda based on the conditions defined by the subscription config:
+- The # of messages in queue is >= `BATCH_SIZE`
+- The time since the last call to the lambda is >= to the `MAX_BATCHING_WINDOW`
+    - Allows us to call the lambda even when the # of messages in the SQS queue is less than the configured `BATCH_SIZE` preventing unnecessary delays.
+
+
+### Option 2 - Leverage Redis as our buffer queue
+Instead of adding a new SQS queue to buffer between the SNS topic and Lambda, Redis is leveraged as the "buffer queue" instead.
+
+1. The lambda remains connected to the SNS topic, and receives receipts 1 by 1 like before
+2. Instead of immediately packaging the receipt as a celery task, the lambda adds the receipt to Redis.
+3. When the # of receipts in Redis >= to the `BATCH_SIZE` the receipts are flushed from Redis, and sent to `process_ses_results` as a celery task similar to the existing implementation
```

Comment:
By the "existing implementation", do we mean the Redis buffer queues where we buffer the notifications at the API level?

Reply:
No, it's referring to how the lambda currently repackages receipts as celery tasks and forwards them to `process_ses_results`. The statement is a bit "fluff" and doesn't add much to the prior description. I'll remove it to avoid confusion.